automatically remove old local peers from peer table

mmckegg opened the pull request:
Currently if sbot discovers a local peer, it will keep it in the peer table forever, even if that local peer goes offline or you change networks. This has a few problems (including sbot trying to connect to missing peers), but also makes it hard to easily access a list of local peers. 

**This PR adds an automatic cleanup to `local` that will remove peers from the `gossip` table that haven't broadcasted in over 10 seconds.**

With this PR in place, `patchwork-next` will no longer have to override the gossip and local modules and will be able to work properly on top of a standard sbot server.

<img width="350" alt="screen shot 2017-02-15 at 4 32 21 pm" src="https://cloud.githubusercontent.com/assets/66834/22959573/5ef1ed1c-f39c-11e6-8505-f5bf314bc695.png">

cc @ahdinosaur @dominictarr @mixmix

mixmix commented:
Code looks clean. I'm not familiar enough with sbot to know how this fits in, but seems like a reasonable change.

Propose fixing those errors Travis is reporting,
Maybe give  @dominictarr a couple of days to comment, then merge

mmckegg commented:
I think the problem is travis is configured with an old node version (iojs-v2.5.0). A test is trying to use `array.find` and throwing an error. This is a fairly new addition to v8 and probably the old node version doesn't support it. /cc @dominictarr 

Here's the travis error:

```
TAP version 13
# alice blocks bob, and bob cannot connect to alice
/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/multiserver/index.js:18
        split(addr).find(function (addr) {
                    ^
TypeError: split(...).find is not a function
    at Object.client (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/multiserver/index.js:18:21)
    at EventEmitter.connect (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/index.js:188:14)
    at EventEmitter.hooked [as connect] (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/hoox/index.js:10:15)
    at /home/travis/build/ssbc/scuttlebot/test/block.js:49:9
    at Object.continuation [as cb] (/home/travis/build/ssbc/scuttlebot/node_modules/cont/node_modules/continuable-para/node_modules/continuable-list/index.js:24:25)
    at Object.cb (/home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/queue.js:13:26)
    at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/util.js:53:12
    at Array.forEach (native)
    at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/util.js:52:14
    at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/level-sublevel/shell.js:92:46
```

mixmix commented:
Can you just change the  .travis.yml

On Sun, 19 Feb 2017 20:09 Matt McKegg <notifications@github.com> wrote:

> I think the problem is travis is configured with an old node version
> (iojs-v2.5.0). A test is trying to use array.find and throwing an error.
> This is a fairly new addition to v8 and probably the old node version
> doesn't support it. /cc @dominictarr <https://github.com/dominictarr>
>
> Here's the travis error:
>
> TAP version 13
> # alice blocks bob, and bob cannot connect to alice
> /home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/multiserver/index.js:18
>         split(addr).find(function (addr) {
>                     ^
> TypeError: split(...).find is not a function
>     at Object.client (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/multiserver/index.js:18:21)
>     at EventEmitter.connect (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/index.js:188:14)
>     at EventEmitter.hooked [as connect] (/home/travis/build/ssbc/scuttlebot/node_modules/secret-stack/node_modules/hoox/index.js:10:15)
>     at /home/travis/build/ssbc/scuttlebot/test/block.js:49:9
>     at Object.continuation [as cb] (/home/travis/build/ssbc/scuttlebot/node_modules/cont/node_modules/continuable-para/node_modules/continuable-list/index.js:24:25)
>     at Object.cb (/home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/queue.js:13:26)
>     at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/util.js:53:12
>     at Array.forEach (native)
>     at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/ssb-feed/util.js:52:14
>     at /home/travis/build/ssbc/scuttlebot/node_modules/secure-scuttlebutt/node_modules/level-sublevel/shell.js:92:46
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ssbc/scuttlebot/pull/373#issuecomment-280901094>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACitnj48l2ADbhXnDYOljc-_jMYvsagrks5rd-q8gaJpZM4MBRH1>
> .
>

mmckegg commented:
Passing now 🎉 !!!

dominictarr commented:
looks great, merging!

dominictarr commented:
merged into 9.4.4, thanks @mmckegg
